### PR TITLE
Support divulged contracts in a distributed ledger

### LINKED ISSUE #2488: Support divulged contracts in a distributed ledger
Divulged contracts in a distributed ledger (i.e. multi-participant topology) are currently not supported. This is witnessed by a foreign key constraint error in the Ledger API Server, when it tries to insert entries into the `contract_divulgences` table.
After a discussion with @oliverse-da and @meiersi-da, we propose the following steps to support divulged contracts:

1. Split the current `contracts` table into `contracts_data` and `contracts`.
    1. `contracts_data` only holds the relation `(ContractId, ContractInstance)`
    1. `contracts` only holds the metadata (i.e. what it currently holds minus the contract instance data)
1. The type of `TransactionAccepted#referenceContracts` is changed to 
`List[(AbsoluteContractId, VersionValue[AbsoluteContractInst)]`
1. `PersistenceEntry.Transaction` is extended with the list of referenced contracts from `TransactionAccepted`.
1. The class `c.d.p.s.s.l.s.d.Contract` needs to change the fields `let`, `transactionId` to be of type `Option`.
1. Stop using `ActiveContract` internally in `ActiveContractsManager`, but only use `Contract`.
1. Rename `LedgerDao#lookupActiveContract` to `LedgerDao#lookupPotentiallyActiveContract`.
1. Only perform the LET contract check if the LET is actually available. For divulged contracts, we can only assume that it is active, as we don't have any further information.
1. When adding contracts coming from a `TransactionAccepted` state update, a "divulged" contract can be upgraded to a disclosed contract by:
    1. checking that the existing `ContractInst` value in `contracts_data` is exactly the same as the to be inserted value from the transaction
    1. additional metadata to the already known entry in `contracts_data` is written to the `contracts` table.
1. A new migration script for postgres must be added to split the `contracts` table into two tables.

Points to think about:
- If a contract is disclosed AFTER it has been divulged (i.e. more metadata like LET and ledger offset is now available), does this have an effect on consistency assumptions? For example, it is possible that `Tx(N).LET < Tx(N-1).LET`, where `Tx(N)` is `Transaction at offset N`.

### PR DESCRIPTION
**Pull request by rautenrieth-da**
Fixes #2488

- Rename ActiveContracts to ActiveLedgerState
- Remove LedgerDao.Contract in favor of ActiveLedgerState.Contract
- Split ActiveLedgerState.Contract into DivulgedContract and ActiveContract
- In JdbcLedgerDao, split the contracts table into contracts and contracts_data
- In InMemoryActiveLedgerState, add a separate map for divulged contracts

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Add a line to the [release notes](https://github.com/digital-asset/daml/blob/master/unreleased.rst), if appropriate
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+      case Some(dc) =>
+        copy(
+          activeContracts = activeContracts + (c.id -> copyDivulgences(c, dc)),
+          divulgedContracts = divulgedContracts - c.id,
+          keys = newKeys
+        )
+    }
+  }
+
+  override def removeContract(cid: AbsoluteContractId, keyO: Option[GlobalKey]) = keyO match {
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala]: This also needs to remove the `cid` from divulged contracts, as a party can exercise a consuming choice on a divulged contract.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveLedgerStateManager.scala [hunk 2]
```diff
@@ -145,10 +157,13 @@ class ActiveContractsManager[ACS](initialState: => ACS)(implicit ACS: ACS => Act
                 ats.copy(
                   errs = contractCheck(absCoid, Exercise).fold(errs)(errs + _),
                   acc = Some(if (ne.consuming) {
-                    acc.removeContract(absCoid, (acc lookupContract absCoid).flatMap(_.key) match {
+                    val keyO = (acc lookupContract absCoid)
+                      .collect({ case c: ActiveContract => c })
+                      .flatMap(_.key) match {
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ActiveLedgerStateManager.scala]: this `match` is really just a `map`.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      // - a divulged contract, in which case its divulgence info is updated
+      // - an unknown contract, in which case a new divulged contract is stored
+      val updatedAcs = activeContracts.intersectWith(global) { (ac, parties) =>
+        ac copy (divulgences = ac.divulgeTo(parties, transactionId))
+      }
+      val updatedDcs = divulgedContracts.intersectWith(global) { (dc, parties) =>
+        dc copy (divulgences = dc.divulgeTo(parties, transactionId))
+      }
+      val newDcs = global.foldLeft(Map.empty[AbsoluteContractId, DivulgedContract])(
+        (m, e) =>
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala]: Please use the partial function syntax to pattern match on `e` and give `_1` and `_2` meaningful names.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+    case None => copy(activeContracts = activeContracts - cid)
+    case Some(key) => copy(activeContracts = activeContracts - cid, keys = keys - key)
+  }
+
+  override def addParties(newParties: Set[Party]): InMemoryActiveLedgerState =
+    copy(parties = newParties.map(p => p -> PartyDetails(p, None, true)).toMap ++ parties)
+
+  override def divulgeAlreadyCommittedContract(
+      transactionId: TransactionIdString,
+      global: Relation[AbsoluteContractId, Party],
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/InMemoryActiveLedgerState.scala]: Please rename `global` (there is no `local`?) to `divulgences` or something in that direction.

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala [hunk 5]
```diff
@@ -87,13 +88,13 @@ class InMemoryLedger(
   // need to take the lock to make sure the two pieces of data are consistent.
   override def snapshot(): Future[LedgerSnapshot] =
     Future.successful(this.synchronized {
-      LedgerSnapshot(entries.ledgerEnd, Source(acs.contracts))
+      LedgerSnapshot(entries.ledgerEnd, Source[ActiveContract](acs.activeContracts.map(_._2)))
```
Comment (gerolf-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala]:
```suggestion
      LedgerSnapshot(entries.ledgerEnd, Source[ActiveContract](acs.activeContracts.values))
```

Comment (rautenrieth-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala]: Now I remember why I had this expression: `acs.activeContracts.values` is a mutable iterable (even though the map is immutable), but `Source` takes an immutable iterable. How do I best convert these? `acs.activeContracts.values.toList`?

Comment (gerolf-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala]: Yikes, that's disappointing of the scala stdlib api. Let's keep it as is.

Comment (rautenrieth-da) [reply ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/inmemory/InMemoryLedger.scala]: Changed to `Source.fromIterator(() => acs.activeContracts.valuesIterator))`

File: ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala [hunk 6]
```diff
@@ -348,7 +370,8 @@ private class JdbcLedgerDao(
       offset: Long,
       tx: Transaction,
       localImplicitDisclosure: Relation[EventId, Party],
-      globalImplicitDisclosure: Relation[AbsoluteContractId, Party])(
+      globalImplicitDisclosure: Relation[AbsoluteContractId, Party],
+      referencedContracts: List[(Value.AbsoluteContractId, AbsoluteContractInst)])(
```
Comment (oliverse-da) [inline ledger/sandbox/src/main/scala/com/digitalasset/platform/sandbox/stores/ledger/sql/dao/JdbcLedgerDao.scala]: Hi Robert - might we worth renaming referencedContracts to divulgedContracts for consistency with the Update event which Jussi recent tweaked.

**Comment by gerolf-da** [comment]
👏 👏 👏

**Comment by oliverse-da** [comment]
Same from me: 👏 🙇

**Pull request merged**